### PR TITLE
fix(front50): save with existing pipeline index

### PR DESCRIPTION
**Pull request by sbwsg**
Pipelines that had been manually sorted in deck would become out of order when roer was used to save over them.  This happened because any existing index property on the saved pipeline was wiped out by the incoming roer pipeline.

This PR checks whether a new pipeline matches application and ID with an existing pipeline and then, if so, copies the existing pipeline's index into the new pipeline so that ordering is maintained.

### CODE REVIEW

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTask.java [hunk 1]
```diff
@@ -99,4 +106,18 @@ public long getBackoffPeriod() {
   public long getTimeout() {
     return TimeUnit.SECONDS.toMillis(30);
   }
+
+  private Map<String, Object> fetchExistingPipeline(Map<String, Object> newPipeline) {
+    String applicationName = (String) newPipeline.get("application");
+    String newPipelineID = (String) newPipeline.get("id");
+    if (newPipelineID != null && !newPipelineID.equals("")) {
```
Comment (jtk54) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTask.java]: Could use `isNotEmpty` etc from [apache StringUtils](https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html).

File: orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTask.java [hunk 2]
```diff
@@ -99,4 +106,18 @@ public long getBackoffPeriod() {
...
+      List<Map<String, Object>> applicationPipelines = front50Service.getPipelines(applicationName);
+      for (Map<String, Object> existingPipeline : applicationPipelines) {
```
Comment (jtk54) [inline orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTask.java]: Could use a java 8 `stream()` with `findFirst()` over a manual loop here.

Comment (lwander) [reply orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTask.java]: The stream api is a little daunting at first:

```java
.stream()
.filter(p -> p.containsKey("id"))
.filter(p -> p.get("id").equals(newPipelineID))
.findFirst().orElse(null)
```

File: orca-front50/src/test/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTaskSpec.groovy [hunk 3]
```diff
@@ -65,4 +65,67 @@ class SavePipelineTaskSpec extends Specification {
     ])
   }
 
+  def "should copy existing pipeline index to new pipeline"() {
+    Integer testIndex = 14
```
Comment (danielpeach) [inline orca-front50/src/test/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTaskSpec.groovy]: From my non-existent knowledge of Spock conventions, I would put this (and `existingPipeline` in `when:`) in the `given:` block. Same for the next spec.

File: orca-front50/src/test/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTaskSpec.groovy [hunk 4]
```diff
@@ -65,4 +65,67 @@ class SavePipelineTaskSpec extends Specification {
...
+    def stage = new Stage(Execution.newPipeline("orca"), "whatever", [
+      pipeline: Base64.encoder.encodeToString(objectMapper.writeValueAsString(pipeline).bytes)
+    ])
+
+    when:
+    Map<String, Object> existingPipeline = [
+      index: testIndex,
+      id: 'existing-pipeline-id',
+    ]
+    Integer index = 0
```
Comment (danielpeach) [inline orca-front50/src/test/groovy/com/netflix/spinnaker/orca/front50/tasks/SavePipelineTaskSpec.groovy]: This spec would read easier to me if you didn't initialize this.

**Pull request merged**